### ISSUE TITLE
Nested structs not handled correctly in codegen

### ISSUE THREAD

**Issue by vezenovm** [noir-lang/noir#529]
# Description

The `codegen_assign` method [here](https://github.com/noir-lang/noir/blob/829c76c04195eb2b5a70c190a90dad9cdb6ecf82/crates/noirc_evaluator/src/ssa/code_gen.rs#L445) has a TODO noting that it is incorrect for nested structs. 

Also when creating an SSA function we flatten the NodeIds for a return value. This includes nested tuple values (which is generalization of both structs and tuples). However, when we set the field member index during type check, this is on an unflattened struct. This means that the when we later fetch a member of the nested struct (which is flattened) we could be fetching a field at the wrong index. 

## Aim

I want to be able to create a nested struct inside of a Noir circuit, and then access that struct's field members accurately later in the circuit.

## Expected behavior

When accessing a struct's field member, I expect it to be the correct field member and type. I should be able to then use this field member in operations with the corresponding types. `nargo build` should also give me an error if I attempt to perform binary operations using mismatched types. 

## Bug

I have a circuit where I generate a nested struct from struct input to main, and assign this struct to a new variable. The parent struct has only Field types, while the nested struct has only integer and boolean types. However, once I try to access a struct field, I find that I am accessing an unexpected field type. 

For example, when trying to multiply a field member of the parent struct that we know to be a `NativeField`, the program panics that we are trying to use a `Boolean` instead. 
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Boolean`,
 right: `NativeField`', crates/noirc_evaluator/src/ssa/node.rs:776:17
```
It is to be noted that is also only error'ing out during evaluation and not during the build stage. This is most likely due to that we correctly match the nested struct fields during our compilation check, however, nested structs are being mishandled during evaluation. If I attempt to directly use the boolean field member I will get an error during nargo build and will never reach the panic in SSA that is posted above, which acts as expected. 

I was able recreate this error when constructing a new struct directly in Noir vs. when using struct inputs from main as the fields of the new struct. I constructed a nested struct directly in Noir, and then reassigned it inside of a `new` function. This led to the same panic. However, when I did not reassign the nested struct through a `new` function, this panic did not happen. This can be seen in the test-case linked below. This is potentially related to issue #492. It seems that this bug may be a result of all nested tuple types being returned from functions incorrectly.

# To reproduce

I have created a test circuit displaying this that can be seen in this PR: https://github.com/vezenovm/aztec-circuits-noir/pull/5. If you run `nargo build --allow-warnings` you will find that the constraint system successfully builds, however, when you run `nargo prove p --allow-warnings` you will get this error:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Unsigned(30)`,
 right: `NativeField`', crates/noirc_evaluator/src/ssa/node.rs:776:17
```

# Environment

I am on the Ubuntu 20.04 mainframe running clang 10.

- OS:

For `nargo` users

- _noir-lang/noir_ commit cloned:
- Proving backend
  - [X] default
    - Clang: (run `clang --version`)
  - [ ] wasm-base

For TypeScript users

- Node.js: (run `node --version`)
- @noir-lang/noir_wasm: (from `yarn.lock`)
- @noir-lang/barretenberg: (from `yarn.lock`)
- @noir-lang/aztec_backend: (from `yarn.lock`)

# Additional context

I have done some brief debugging, and the generation of witnesses for a struct as part of the Abi looks to be correct, however, further debugging might have to be done here. As I am currently only able to reproduce this error in the example linked above, it is worthy to check whether this error is unique to struct ABI inputs. 

I have a suspicion though it is an error when trying to use a nested struct at all in an assignment as stated by the TODO in `codegen_assign`, or an error in returning nested structs from functions.

**Comment by okuyiga** [comment]
Would this also help with https://github.com/noir-lang/noir/issues/492 ?

**Comment by vezenovm** [comment]
Thanks for pointing that out. I thought those issues may be related as I am only getting the error when returning the nested struct from a method. I will determine whether they are related.

**Comment by vezenovm** [comment]
Issue solved by #542
